Add option to capture mouse

HifiExperiments opened the pull request:
(Currently not hooked up to any scripts or keys)

To enable, in console: `Camera.captureMouse = true;`

Set to false to disable.

Needs to be tested on other OSes, and not sure how it interacts with the mouse logic in HMD

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-b60ae9f_1403f890.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-b60ae9f.tar_1403f890.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-b60ae9f.tar_1403f890.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-b60ae9f_1403f890.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-b60ae9f_1403f890.exe

vegaslon commented:
Here is a quck script made to test it, uses the "m" key to enable and disable.
https://gist.githubusercontent.com/vegaslon/0bf901ed5cf2fdcdddded61dbcdd5ded/raw/1e2cbbc41fbd3b846ebc699f22e00a2f87a44a61/Mousecapturewithm.js

vegaslon commented:
Did not see it cause any trouble when enabled and disable in VR mode with a Occulus Rift. The was functional again when disabled.

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +111,20 @@ public slots:
+    /**jsdoc
+     * Sets mouse capture state.
+     * @function Camera.setCaptureMouse
+     * @param {boolean} captureMouse - Whether or not to capture the mouse.
```
```suggestion
     * @param {boolean} captureMouse - <code>true</code> to capture the mouse, <code>false</code> to release the mouse.
```

ctrlaltdavid commented:
Also, update copyrights in:
- CompositorHelper.cpp
- KeyboardMouseDevice.cpp
- KeyboardMouseDevice.h
- Camera.h

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-e19a419.tar_932f5dbf.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-e19a419_932f5dbf.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-e19a419.tar_932f5dbf.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-e19a419_932f5dbf.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-e19a419_932f5dbf.exe

JulianGro commented:
Mouse look works on Linux Mint 19.3, but its performance is pretty bad (same as when holding right click). It is very sensitive and kind of makes jumps, as if it took the mouse moving a pixel and multiplied it with 10 to move 10 pixel at once. Even on the lowest sensitivity setting of my mouse, the movement is not smooth at all.
This jumpy hypersensitivity constantly makes me lose orientation.

HifiExperiments commented:
@JulianGro interesting!  what’s your framerate like?  it sounds like it’s not a new problem so I wouldn’t block this PR, but we can add mouse sensitivity settings or make the camera framerate independent if it’s not already.  it’s weird because all the calculations are just based on pixel differences, so I’m not sure why linux would be different or why framerate would affect it

JulianGro commented:
After you asking, I tried it in the serverless rather than the hub, and it does feel smooth in there. The sensitivity is still stupidly high though. I measured it and a 360° spin is just 3mm of mouse movement. For comparison, moving the cursor from one side of the screen to the other is 1cm of mouse movement.
Also I just noticed that the losing orientation in the hub isn't actually because of the high sensitivity, but because the speed is not consistent there. It seems like the movement speeds up when the "gamerate" dips down.

ctrlaltdavid commented on interface/src/Application.cpp:
```diff
@@ -680,6 +680,7 @@ class ApplicationMeshProvider : public scriptable::ModelProviderFactory  {
  *     <tr><td><code>CameraIndependent</code></td><td>number</td><td>number</td><td>The camera is in independent mode.</td></tr>
  *     <tr><td><code>CameraEntity</code></td><td>number</td><td>number</td><td>The camera is in entity mode.</td></tr>
  *     <tr><td><code>InHMD</code></td><td>number</td><td>number</td><td>The user is in HMD mode.</td></tr>
+ *     <tr><td><code>CaptureMouse</code></td><td>number</td><td>number</td><td>The mouse is captured.</td></tr>
```
Expand on the description - what does it mean for the mouse to be "captured"?

ctrlaltdavid commented on libraries/input-plugins/src/input-plugins/KeyboardMouseDevice.cpp:
```diff
@@ -113,9 +113,16 @@ void KeyboardMouseDevice::eraseMouseClicked() {
+        _accumulatedMove += currentPos - _lastCursor;
+    } else if (!isNaN(captureTarget.x())) {
+        QPointF change = event->globalPos() - captureTarget;
+        _accumulatedMove += QPoint(change.x(), change.y());
```
If `mouseMoveEvent()` calls happen more often than calls to `pluginUpdate()` then the amount of movement will be multiplied?

HifiExperiments replied:
not sure what you mean.  it’s additive, so this _accumulatedMove will get larger if multiple mouse moves happen at once.  perhaps this explains the framerate dependent behavior that Julian is seeing, but it’s equivalent to how it’s done in master now (the difference is just calculated here instead of above)

ctrlaltdavid replied:
OK

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +112,20 @@ public slots:
     void setOrientation(const glm::quat& orientation);
 
+    /**jsdoc
+     * Gets the current mouse capture state.
```
What does it mean for the mouse to be "captured"?

HifiExperiments commented on interface/src/Application.cpp:
```diff
@@ -715,6 +716,7 @@ static const QString STATE_PLATFORM_ANDROID = "PlatformAndroid";
 static const QString STATE_LEFT_HAND_DOMINANT = "LeftHandDominant";
 static const QString STATE_RIGHT_HAND_DOMINANT = "RightHandDominant";
 static const QString STATE_STRAFE_ENABLED = "StrafeEnabled";
+static const QString STATE_CAMERA_CAPTURE_MOUSE = "CaptureMouse";
```
It’s not quite a camera mode (in that you can be in first person or third person or otherwise and have the mouse captured or not) but it does affect the camera...I will change to STATE_CAPTURE_MOUSE

JulianGro commented:
If this can apparently enabled by just any script, it should definitely be overwritten by something like the escape key.
Also I don't know if we have much in terms of security for scripts, but even with an override, I don't think there is anything that would stop someone from putting a script on their avatar that constantly captures and frees the cursor to potentially make it impossible to open the menu to block/mute/kick/ban them.

HifiExperiments commented:
@JulianGro "capturing" the mouse in this way is actually already possible via Reticle.visible and Reticle.position, which similarly hide the mouse and can take control of it.  this new mode has the advantage that it relinquishes control when Interface is not active or when a menu item is selected, which allows you to get free (like the mouse capture in HMD).  it is true that we'll need a key to get in/out of this mode by default (the issue suggested Alt) and escape (probably via away.js) should always exit this mode and then restore it when you come back, but I've left those to be designed for another PR (and maybe someone who scripts more than me).

and then of course there's a larger discussion about script security + protecting against malicious scripts, but that affects pretty much every aspect of the scripting

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-894791e.tar_9546cebe.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-894791e.tar_9546cebe.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-894791e_9546cebe.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-894791e_9546cebe.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-894791e_9546cebe.dmg

HifiExperiments commented:
@kasenvr @ctrlaltdavid @JulianGro I've added an additional `Camera.sensitivity` setting so that people have some more control.  This doesn't fix the framerate issue Julian mentioned, it just allows people to dial the speed of the camera rotation (whether the mouse is captured or not).  I added a slider for this value to Settings -> Controls.

I'm noticing that these preferences might not be saved, which is annoying...but instead of handling it for just this value, I think we should fix that for everything (and do a general push to clean up the settings and expose more helpful things)

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-895d013.tar_e4f32f1c.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-895d013_e4f32f1c.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-895d013.tar_e4f32f1c.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-895d013_e4f32f1c.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-895d013_e4f32f1c.exe

kasenvr commented:
I thought the Settings API was pretty safe in terms of saving and retrieving. Is it no longer working as expected or which part of things is giving trouble, exactly?

HifiExperiments commented:
the settings API works, but the Preferences dialogs don’t seem to be all backed by settings, which is why they aren’t saved

ctrlaltdavid commented on interface/resources/controllers/keyboardMouse.json:
```diff
@@ -72,156 +72,86 @@
           "filters":
             [
-                { "type": "scale", "scale": 0.6 }
+                { "type": "scale", "scale": 0.2 }
```
Is this change of scale value intentional? Does it adversely affect current sensitivity / speed of movement?

HifiExperiments replied:
This is a matter of the diff being confusing.  The block above this handles the old cases when !Application.CaptureMouse, and still has scale 0.6.  because of how the mouse capture mode works, we want this to be less sensitive.  then on top of that, you can now control the sensitivity further with Camera.sensitivity (with the caveat of framerate-dependent behavior as discussed previously)

ctrlaltdavid commented on interface/resources/controllers/keyboardMouse.json:
```diff
@@ -72,156 +72,86 @@
+          "when": "Application.CaptureMouse",
+          "filters":
+            [
+                { "type": "scale", "scale": 0.2 }
```
Ditto.

HifiExperiments replied:
same here, this is the new capture case where we want lower sensitivity

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -43,6 +44,8 @@ class Camera : public QObject {
     Q_PROPERTY(glm::quat orientation READ getOrientation WRITE setOrientation)
     Q_PROPERTY(QString mode READ getModeString WRITE setModeString NOTIFY modeUpdated)
     Q_PROPERTY(QVariantMap frustum READ getViewFrustum CONSTANT)
+    Q_PROPERTY(bool captureMouse READ getCaptureMouse WRITE setCaptureMouse NOTIFY captureMouseUpdated)
+    Q_PROPERTY(float sensitivity READ getSensitivity WRITE setSensitivity)
```
JSDoc for these two new properties needs to be added to FancyCamera.h.

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +113,36 @@ public slots:
+    bool getCaptureMouse() const { return _captureMouse; }
+
+    /**jsdoc
+     * Sets the mouse capture state.  When <code>true</code>, the mouse will be invisible and cannot leave the bounds of
+     * Interface, as long as Interface is the active window and no menu item is selected.  When <code>false</code>, the mouse
+     * will behave normally.
```
```suggestion
     * Sets the mouse capture state.  When <code>true</code>, the mouse is invisible and cannot leave the bounds of
     * Interface, as long as Interface is the active window and no menu item is selected.  When <code>false</code>, the mouse
     * behaves normally.
```

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +113,36 @@ public slots:
 
+    /**jsdoc
+     * Gets the current mouse capture state.
+     * @function Camera.getMouseCapture
+     * @returns {boolean} The current mouse capture state.
```
```suggestion
     * @function Camera.getCaptureMouse
     * @returns {boolean} <code>true</code> if the mouse is captured (is invisible and cannot leave the bounds of Interface, if Interface is the active window and no menu item is selected), <code>false</code> if the mouse is behaving normally.
```

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +113,36 @@ public slots:
+    /**jsdoc
+     * Gets the current camera sensitivity.
+     * @function Camera.getSensitivity
+     * @returns {boolean} The current camera sensitivity.
```
```suggestion
     * @returns {number} The current camera sensitivity. TODO: Describe the permissible range of values.
```

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +113,36 @@ public slots:
+    /**jsdoc
+     * Sets the camera sensitivity.  Higher values mean that the camera will be more sensitive to mouse movements.
+     * @function Camera.setSensitivity
+     * @param {boolean} sensitivity - The desired camera sensitivity.
```
```suggestion
     * @param {number} sensitivity - The desired camera sensitivity. TODO: Describe the permissible range of values.
```

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -110,6 +113,36 @@ public slots:
+     * @function Camera.setSensitivity
+     * @param {boolean} sensitivity - The desired camera sensitivity.
+     */
+    void setSensitivity(float sensitivity) { _sensitivity = glm::max(0.0f, sensitivity); }
```
Should there be an upper bound to the sensitivity value?

HifiExperiments replied:
I think not needed for now.  In the Preferences dialog it maxes out at 5, but no need to restrict people

ctrlaltdavid commented on libraries/shared/src/shared/Camera.h:
```diff
@@ -181,6 +214,20 @@ public slots:
+     *
+     * Camera.captureMouseUpdated.connect(onCaptureMouseUpdated);
+     */
+    void captureMouseUpdated(bool newCaptureMouse);
```
```suggestion
    void captureMouseChanged(bool newCaptureMouse);
```
`captureMouseChanged()` would be a better name for this signal. (Most such signals use "Changed" rather than "Updated".)

ctrlaltdavid reviewed (changes requested):
Other items in the PreferencesDialog.cpp have their settings saved - in a round about manner. For example, "analog plus walk speed" setting is backed by `Setting::Handle<float> _analogPlusWalkSpeedSetting;` and the setting is saved in `MyAvatar::saveData()`.

kasenvr commented:
Will give this a QA pass after the CR concludes, then I'll have Nova also test it (Windows) to confirm that all's well for the first-pass implementation then we'll be done. Quick and easy. :)

vircadia-build-notifier commented:
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-1a11d6c.tar_e0289436.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-1a11d6c.tar_e0289436.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-1a11d6c_e0289436.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-1a11d6c_e0289436.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR677/Vircadia-Alpha-PR677-1a11d6c_e0289436.exe

kasenvr reviewed (approved):
Appears to work as advertised. One note is that the sensitivity setting seems to be in an odd place, it should get its own area instead of being under "VR Movement", perhaps a "Desktop" section at the top?

Also a very lovely surprise is that this works great for TPS! Maybe we'll want horizontal and vertical sensitivity later + acceleration options, but for now this is a perfect start. Thanks so much for your help on this! I know it was tough but I think it's well worth it. :)

vegaslon commented:
Tested and seems to work as planned, although at really slow sensitivity the mouse cursor is seen flickering on the screen. That is no reason to not merge this PR just an observation.
